File: slideo/timeselectdialog.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

from PyQt5.QtWidgets import QWidget, QDialog, QTimeEdit, QPushButton, QFormLayout, QVBoxLayout, QHBoxLayout
from PyQt5.QtCore import QTime

class TimeSelectDialog(QDialog):
    def __init__(self, parent, videoManager, windowTitle, timeSelectorLabel):
        super().__init__(parent)

        self.videoManager = videoManager

        position = videoManager.player.position()
        qPosition = QTime(0, 0, 0).addMSecs(position)

        mainLayout = QVBoxLayout()

        formLayout = QFormLayout()
        self.timeEditor = QTimeEdit(qPosition)
        self.timeEditor.setDisplayFormat("HH:mm:ss.zzz")
        formLayout.addRow(timeSelectorLabel, self.timeEditor)

        formWidget = QWidget()
        formWidget.setLayout(formLayout)

        buttonsLayout = QHBoxLayout()

        buttonsLayout.addStretch()

        self.cancelButton = QPushButton("Cancel")
        buttonsLayout.addWidget(self.cancelButton)

        self.okButton = QPushButton("OK")
        self.okButton.setDefault(True)
        buttonsLayout.addWidget(self.okButton)

        buttonsWidget = QWidget()
        buttonsWidget.setLayout(buttonsLayout)

        mainLayout.addWidget(formWidget)
        mainLayout.addWidget(buttonsWidget)

        self.setLayout(mainLayout)
        self.setWindowTitle(windowTitle)

class JumpToTimeDialog(TimeSelectDialog):
    def __init__(self, parent, videoManager):
        super().__init__(parent, videoManager, "Jump to time", "Time to jump to:")

        self.cancelButton.clicked.connect(self.cancel)
        self.okButton.clicked.connect(self.validate)
        self.timeEditor.editingFinished.connect(self.validate)

    def cancel(self):
        self.done(1)

    def validate(self):
        self.videoManager.player.setPosition(-self.timeEditor.dateTime().time().msecsTo(QTime(0, 0, 0)))
        self.done(0)

class AddBreakPointDialog(TimeSelectDialog):
    def __init__(self, parent, videoManager, projectManager):
        super().__init__(parent, videoManager, "Add breakpoint", "Breakpoint position:")

        self.projectManager = projectManager

        self.cancelButton.clicked.connect(self.cancel)
        self.okButton.clicked.connect(self.validate)
        self.timeEditor.editingFinished.connect(self.validate)

    def cancel(self):
        self.done(1)

    def validate(self):
        self.projectManager.addBreakpoint(-self.timeEditor.dateTime().time().msecsTo(QTime(0, 0, 0)))
        self.done(0)

File: src/timeselectdialog.cpp
#include "timeselectdialog.hpp"

#include "mainwindow.hpp"

#include <QVBoxLayout>
#include <QHBoxLayout>
#include <QFormLayout>

#include <QTime>

#include <QEvent>
#include <QKeyEvent>

#include <QApplication>

TimeSelectDialog::TimeSelectDialog(QWidget& parent, QString windowTitle, QString timeSelectorLabel)
      : QDialog(&parent)
      , parent(parent)
      , timeEditor()
      , cancelButton("Cancel")
      , validateButton("OK") {
	MainWindow& mwParent = dynamic_cast<MainWindow&>(parent);

	QTime currentPosition = QTime(0, 0, 0, 0).addMSecs(mwParent.getVideoPlayer().getPosition());

	QVBoxLayout* mainLayout = new QVBoxLayout;

	QFormLayout* formLayout = new QFormLayout;
	timeEditor.setTime(currentPosition);
	timeEditor.setDisplayFormat("HH:mm:ss.zzz");
	formLayout->addRow(timeSelectorLabel, &timeEditor);

	QWidget* formWidget = new QWidget;
	formWidget->setLayout(formLayout);

	QHBoxLayout* buttonsLayout = new QHBoxLayout;
	buttonsLayout->addStretch();
	buttonsLayout->addWidget(&cancelButton);
	buttonsLayout->addWidget(&validateButton);

	validateButton.setDefault(true);

	QWidget* buttonsWidget = new QWidget;
	buttonsWidget->setLayout(buttonsLayout);

	mainLayout->addWidget(formWidget);
	mainLayout->addWidget(buttonsWidget);

	setLayout(mainLayout);
	setWindowTitle(windowTitle);

	connect(&cancelButton, SIGNAL(clicked()), this, SLOT(cancel()));
	connect(&validateButton, SIGNAL(clicked()), this, SLOT(validate()));

	qApp->installEventFilter(this);
}

void TimeSelectDialog::cancel() {
	done(1);
}

qint64 TimeSelectDialog::getTime() const {
	return QTime(0, 0, 0, 0).msecsTo(timeEditor.dateTime().time());
}

bool TimeSelectDialog::eventFilter(QObject* obj, QEvent* event) {
	if(obj == &timeEditor && event->type() == QEvent::KeyPress) {
		QKeyEvent *keyEvent = dynamic_cast<QKeyEvent*>(event);
		if(keyEvent->key() == Qt::Key_Enter || keyEvent->key() == Qt::Key_Return) {
			validate();
			return true;
		}
	}
	return false;
}

JumpToTimeDialog::JumpToTimeDialog(QWidget& parent)
      : TimeSelectDialog(parent, "Jump to time", "Time to jump to:") {}

void JumpToTimeDialog::validate() {
	MainWindow& mwParent = dynamic_cast<MainWindow&>(parent);
	mwParent.setVideoPlayerPosition(QTime(0, 0, 0, 0).msecsTo(timeEditor.dateTime().time()));
	done(0);
}

AddBreakpointDialog::AddBreakpointDialog(QWidget& parent)
      : TimeSelectDialog(parent, "Add breakpoint", "Breakpoint position:") {}

void AddBreakpointDialog::validate() {
	MainWindow& mwParent = dynamic_cast<MainWindow&>(parent);
	mwParent.addProjectBreakpoint(getTime());
	done(0);
}
